### ISSUE TITLE
tile_to_layout, reproject, pyramid layout consistency

### ISSUE THREAD

**Issue by echeipesh** [locationtech-labs/geopyspark#369]
The core issue is that there are two ways to tile a layer in GeoTrellis parlance:
- "floating" snapping layout to the enveloper raster pixels, matching its resolution
- "zoom" snapping layout to world extent, matching resolution in one of the layers of power of 2 pyramid.

When working with layers there is a lot of benefit to being in "floating" layout when displaying them you need to be in the "zoom" layout. Current state of API makes these transitions hidden and painful. Not to mention that nomenclature is not laden with meaning.

The solution is to overload the three methods that need to deal with layouts to be helpful and hide the mechanics of the changes:

Reading GeoTiffs and tiling them
```py
elevation = geotiff.get("s3://blah/blah/blah.tif")

# match the layout of another layer, hope CRS line up
elevation.tile_to_layout(layout=nlcd)

# match layout of another metadata, effectively making a study area
elevation.tile_to_layout(layout=StudyAreaMetadata)

# Collect extent of the tiffs and snap extent to their envelope in collected CRS
elevation.tile_to_layout(layout=LocalLayout(tile_size=512)) 

# Collect extent of the tiffs and snap extent to world extent in collected CRS
elevation.tile_to_layout(layout=GlobalLayout(tile_size=512))

# Same as above but compensate for missing CRS tiff tags
elevation.tile_to_layout(layout=GlobalLayout(tile_size=512), force_crs="epsg:3857")
```

Reproject a tiled layer
```
def reproject(
    self, 
    target_crs,
    layout=None,
    resample_method=ResampleMethod.NEAREST_NEIGHBOR
)


# reproject and match an existing layout, which is hopefully in the same CRS
elevation.reproject("epsg:3857", layout=nlcd)

# Preserve layout as much as reasonable, infer if its float of power of 2 based on layout
layer.reproject("epsg:3857", layout=None)

# reproject and keep resolution after reprojectiong, using floating layout scheme
layer.reproject("epsg:3857", layout=LocalLayout(tile_size=256))

# reproject and fit to global layout, inferring new zoom level
layer.reproject("epsg:3857", layout=GlobalLayout())
```

Time to pyramid, maybe for display?
```py
# pyramid up until you get to single tile, whatever that means
elevation_pyramid = elevation.pyramid()
server = TMSErver(elevation_pyramid)
# Error: not a global layout, PC LOAD LETTER!

# Ok, we can fix that easily like so ....
elevation_pyramid = elevation.tile_to_layout(GlobalLayout()).pyramid()
server = TMSServer(elevation_pyramid)
# YES! Pixels on a map!
```

Key things that fall out of this:
- `GlobalLayout` and `LocalLayout` are more intuitive than "float" and "Zoom"
- `GlobalLayout` and `LocalLayout` used in `RasterLayer.tile_to_layout` require a collect metadata
- `GlobalLayout` and `LocalLayout` used in `TiledLayer.tile_to_layout` use layer metadata to produce new layouts
- `reproject` has a default setting that does the right thing
- Using `Metadata` instance of `TiledRaterLayer` instance in place of layout matches their layouts so operations can be performed easily.
- `pyramid` does not have to switch convention from defaulting to `LocalLayout` to `GlobalLayout` because the error can be clear and very easy to correct
- `TMSServer` may in the future be able to display `LocalLayout`s with on the fly resampling, removing the error without surprise
- `RasterLayer.to_tiled_layer` is not needed because `tile_to_layout` handles its job more consistently.

After a little investigation its clear that `TiledLayerMetadata.fromRDD` is unhelpful in this use case because it makes an assumption that somehow the CRS will be known or enforced, implying immediate followup `reproject`. That will need to be re-implemented in helpful manner in GeoPySpark.

**Comment by echeipesh** [comment]
Consider: a function that re-tiles and re-projects in a single step.
Use case: Re tile to web mercator global layout as as one step required to display a layer on TMS

**Comment by jbouffard** [comment]
Resolved in #379
